File: ServerLib/LogUpload/src/main/java/com/yc/logupload/request/RequestManager.java
package com.yc.logupload.request;

import com.yc.logupload.config.UploadInitHelper;
import com.yc.logupload.inter.OnUploadListener;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.net.FileNameMap;
import java.net.URLConnection;
import java.util.concurrent.TimeUnit;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class RequestManager {

    private static final OkHttpClient sHttpClient;
    private static volatile RequestManager sInstance;

    static {
        sHttpClient = new OkHttpClient.Builder()
                .connectTimeout(150, TimeUnit.SECONDS)
                .readTimeout(150, TimeUnit.SECONDS)
                .readTimeout(150, TimeUnit.SECONDS)
                .build();
    }

    private RequestManager() {

    }

    public static RequestManager getInstance() {
        if (sInstance == null) {
            synchronized (RequestManager.class) {
                if (sInstance == null) {
                    sInstance = new RequestManager();
                }
            }
        }
        return sInstance;
    }

    /**
     * 上传文件到服务器
     * @param file                  file文件
     * @param onUploadListener      上传回调
     */
    public void uploadFile(File file, final OnUploadListener onUploadListener) {
        String serverHost = UploadInitHelper.getConfig().getServerUrl();
        if (serverHost!=null){
            if (!serverHost.startsWith("http") || !serverHost.startsWith("https")) {
                onUploadListener.onError("url must be contain http or https");
                return;
            }
            postUploadFile(serverHost, file.getPath(), file.getName(), new Callback() {
                @Override
                public void onFailure(@NotNull Call call, @NotNull IOException e) {
                    onUploadListener.onError("post net error :" + e.getMessage());
                }

                @Override
                public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {
                    onUploadListener.onSuccess();
                }
            });
        }
    }


    /**
     * 上传文件，post请求
     * @param url               请求地址
     * @param pathName          文件路径名称
     * @param fileName          文件名
     * @param callback          callback
     */
    public static void postUploadFile(String url, String pathName,
                                      String fileName, Callback callback) {
        //判断文件类型
        MediaType mediaType = MediaType.parse(judgeType(pathName));
        //创建文件参数
        MultipartBody.Builder builder ;
        if (mediaType != null) {
            RequestBody requestBody = RequestBody.create(mediaType, new File(pathName));
            builder = new MultipartBody.Builder()
                    .setType(MultipartBody.FORM)
                    .addFormDataPart(mediaType.type(), fileName, requestBody);
            //发出请求参数
            Request request = new Request.Builder()
                    .url(url)
                    .post(builder.build())
                    .build();
            Call call = sHttpClient.newCall(request);
            call.enqueue(callback);
        }
    }

    /**
     * 根据文件路径判断MediaType
     * @param path              文件路径
     * @return                  返回字符串类型的文件类型
     */
    private static String judgeType(String path) {
        FileNameMap fileNameMap = URLConnection.getFileNameMap();
        String contentTypeFor = fileNameMap.getContentTypeFor(path);
        if (contentTypeFor == null) {
            contentTypeFor = "application/octet-stream";
        }
        return contentTypeFor;
    }


}

File: ServerLib/LogUpload/src/main/java/com/yc/logupload/request/RequestManager.kt
package com.yc.logupload.request

import android.widget.Toast
import com.yc.logupload.config.UploadInitHelper
import com.yc.logupload.inter.OnUploadListener
import com.yc.activitymanager.ActivityManager
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import java.io.File
import java.io.IOException
import java.net.URLConnection
import java.util.concurrent.TimeUnit

/**
 * @author yangchong
 * email  : yangchong211@163.com
 * time   : 2022/5/11
 * desc   : 网络请求管理类
 * revise :
 */
class RequestManager private constructor() {

    companion object {

        private var sHttpClient: OkHttpClient? = null

        init {
            sHttpClient = OkHttpClient.Builder()
                .connectTimeout(150, TimeUnit.SECONDS)
                .readTimeout(150, TimeUnit.SECONDS)
                .readTimeout(150, TimeUnit.SECONDS)
                .build()
        }

        @Volatile
        private var sInstance: RequestManager? = null
        val instance: RequestManager?
            get() {
                if (sInstance == null) {
                    synchronized(RequestManager::class.java) {
                        if (sInstance == null) {
                            sInstance = RequestManager()
                        }
                    }
                }
                return sInstance
            }

        /**
         * 上传文件，post请求
         * @param url               请求地址
         * @param pathName          文件路径名称
         * @param fileName          文件名
         * @param callback          callback
         */
        fun postUploadFile(url: String?, pathName: String, fileName: String?, callback: Callback?) {
            //判断文件类型
            val judgeType = judgeType(pathName)
            val mediaType = judgeType.toMediaTypeOrNull()
            //创建文件参数
            val builder: MultipartBody.Builder
            if (mediaType != null) {
                builder = MultipartBody.Builder()
                    .setType(MultipartBody.FORM)
                    .addFormDataPart(
                        mediaType.type, fileName,
                        RequestBody.create(mediaType, File(pathName))
                    )
                //发出请求参数
                val request: Request = Request.Builder()
                    .url(url.toString())
                    .post(builder.build())
                    .build()
                val call = sHttpClient?.newCall(request)
                call?.enqueue(callback!!)
            }
        }

        /**
         * 根据文件路径判断MediaType
         * @param path              文件路径
         * @return                  返回字符串类型的文件类型
         */
        private fun judgeType(path: String): String {
            val fileNameMap = URLConnection.getFileNameMap()
            var contentTypeFor = fileNameMap.getContentTypeFor(path)
            if (contentTypeFor == null) {
                contentTypeFor = "application/octet-stream"
            }
            return contentTypeFor
        }
    }

    /**
     * 上传文件到服务器
     * @param file                  file文件
     * @param onUploadListener      上传回调
     */
    fun uploadFile(file: File?, onUploadListener: OnUploadListener?) {
        val url = UploadInitHelper.config?.serverUrl
        if (url != null && file != null) {
            if (!url.startsWith("http") || !url.startsWith("https")) {
                onUploadListener?.onError("url must be contain http or https")
                return
            }
            val activity = ActivityManager.getInstance().peek()
            postUploadFile(url, file.path, file.name, object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    onUploadListener?.onError("post net error :" + e.message)
                    activity?.runOnUiThread {
                        Toast.makeText(activity,"上传失败",Toast.LENGTH_SHORT).show()
                    }
                }

                override fun onResponse(call: Call, response: Response) {
                    onUploadListener?.onSuccess()
                    activity?.runOnUiThread {
                        Toast.makeText(activity,"上传成功",Toast.LENGTH_SHORT).show()
                    }
                }

            })
        } else {
            onUploadListener?.onError("url or file is null")
        }
    }

}